### DISCUSSION TITLE
Can't seem to make borders work on empty cells

### DISCUSSION THREAD

**Discussion by mferino** [exceljs/exceljs#1755] [Q&A] [answered]
It appears this has been asked in the past and the answer seems to be including empty string in each cell that is doesn't otherwise have data in it but that's going to require me to refactor a lot of code in order to accomplish this.  Basically when I do something like this ->
```
worksheet.getRow(6).eachCell({ includeEmpty: true }, (cell, colNumber) => {
    cell.border = thinBorderAllSides;
});
```
it breaks down whenever there is an empty cell and so I can't style the entire row. I noticed that someone suggested applying the styles to the entire row instead of each cell but that had no effect either.  Thanks for the help!

**Comment by mferino** [comment, marked as answer]
For those in the future that have this issue and can't seem to find a solution I found (in my mind) a better workaround than adding the empty string to each empty cell.  The way around this is that if you use worksheet.getCell('cellAddress') and apply styles to that cell, it works as expected.  Essentially all you have to know is the number of cells in each row as `totalCells` - I used this function to convert numbers to cell sheet letters
```js
const numberToLetters = (num) => {
            let letters = "";
            while (num >= 0) {
                letters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"[num % 26] + letters;
                num = Math.floor(num / 26) - 1;
            }
            return letters;
        };
```
then I used this function to iterate over all the cells and apply the styling
```js
const styleARowByCell = (totalCells, styleObject, rowNum) => {
            for (let i = 0; i < totalCells; i++) {
                const cellAddr = `${numberToLetters(i)}${rowNum}`;
                const cell = worksheet.getCell(cellAddr);
                for (const [key, value] of Object.entries(styleObject)) {
                    cell[key] = value;
                }
            }
        };
```
and I just passed in an object with key value pairs of the styling I wanted to provide to each cell like so ->
```js
const styleObj = {
            font: { size: 12 },
            border: thinBorderAllSides,
            alignment: midLeftText,
        };
```
and although it's more verbose in the styling code, I didn't have to change how the program that built the data and piped it into the spreadsheet worked.  Hope this is helpful to someone!

**Reply by Siemienik** [reply]
1. ExcelJS has `Range. forEachAddress()`. I suppose that may make your code simpler ;)

https://github.com/exceljs/exceljs/blob/331602b6891457ea4d74385fe1d5466c459fae1f/lib/doc/range.js#L248-L254

2. Cell has property `style`, 

https://github.com/exceljs/exceljs/blob/331602b6891457ea4d74385fe1d5466c459fae1f/lib/doc/cell.js#L27

_It is recommended to make a deep copy before setting styles (eg. `cell.styles = JSON.parse(JSON.stringify(myStyles))_
